File: python/Windows/shared/shared_utils/pixyz_Utils/pixyz_utils.py
import time
from pxz import core, scene, view, geom
from enum import Enum
import json


def getStats(root):
    core.configureInterfaceLogger(False, False, False)  # hide next lines from logs

    t = time.time()
    n_triangles = scene.getPolygonCount([root], True, False, False)
    n_vertices = scene.getVertexCount([root], False, False, False)
    n_parts = len(scene.getPartOccurrences(root))

    core.configureInterfaceLogger(True, True, True)  # reenable logs

    return t, n_triangles, n_vertices, n_parts


class Orientation(Enum):
    FRONT = 0
    BACK = 1
    TOP = 2
    BOTTOM = 3
    LEFT = 4
    RIGHT = 5


def save_screenshot(occurrence, path, orientation=Orientation.FRONT, type=view.CameraType.Perspective, resolution=1024,
                    fov=60, show_edges=False, show_lines=False):
    print("Taking Screenshot...")
    if orientation == Orientation.FRONT:
        direction = geom.Point3(0, 0, -1)
    elif orientation == Orientation.BACK:
        direction = geom.Point3(0, 0, 1)
    elif orientation == Orientation.TOP:
        direction = geom.Point3(0, -1, 0)
    elif orientation == Orientation.BOTTOM:
        direction = geom.Point3(0, 1, 0)
    elif orientation == Orientation.LEFT:
        direction = geom.Point3(1, 0, 0)
    elif orientation == Orientation.RIGHT:
        direction = geom.Point3(-1, 0, 0)
    else:
        direction = geom.Point3(0, 0, 0)

    viewer = view.createViewer(resolution, resolution)
    gpu_scene = view.createGPUScene([occurrence], show_edges)
    view.addGPUScene(gpu_scene, viewer)
    view.fitCamera(direction, type, fov, viewer, occurrence)
    if show_edges:
        view.setViewerProperty("ShowEdges", "True", viewer)
    if show_lines:
        view.setViewerProperty("ShowLines", "False", viewer)

    view.takeScreenshot(path, viewer)
    view.destroyViewer(viewer)
    view.destroyGPUScene(gpu_scene)


def printStats(fileName, t, n_triangles, _n_triangles, n_vertices, _n_vertices, n_parts, _n_parts):
    print('\n')
    print('{:<20s}{:<3s}\n'.format('file ', fileName))
    print('{:<20s}{:<8.3f}{:<3s}\n'.format('optimization ', t, ' s'))
    print('{:<20s}{:<3s}\n'.format('triangles ', str(n_triangles) + ' -> ' + str(_n_triangles)))
    print('{:<20s}{:<3s}\n'.format('vertices ', str(n_vertices) + ' -> ' + str(_n_vertices)))
    print('{:<20s}{:<3s}\n'.format('parts ', str(n_parts) + ' -> ' + str(_n_parts)))


def get_metadata_dict(metadata):
    metadata_dict = dict()
    metadata_dict['name'] = metadata.name
    metadata_dict['value'] = metadata.value
    return metadata_dict


def get_hierarchy_dict(occurrence):
    hierarchy_dict = dict()
    hierarchy_dict['name'] = core.getProperty(occurrence, "Name")
    matrix = scene.getLocalMatrix(occurrence)
    vector3lst = geom.toTRS(matrix)
    hierarchy_dict['translation'] = str([vector3lst[0].x, vector3lst[0].y,vector3lst[0].z])
    hierarchy_dict['rotation'] = str([vector3lst[1].x, vector3lst[1].y,vector3lst[1].z])
    hierarchy_dict['scale'] = str([vector3lst[2].x, vector3lst[2].y,vector3lst[2].z])
    hierarchy_dict['children'] = [get_hierarchy_dict(child) for child in scene.getChildren(occurrence)]
    if scene.hasComponent(occurrence, scene.ComponentType.Metadata):
        metadata = scene.getComponent(occurrence, scene.ComponentType.Metadata)
        metadata_definitions = scene.getMetadatasDefinitions([metadata])[0]
        hierarchy_dict['metadata'] = [get_metadata_dict(metadata_definition) for metadata_definition in
                                      metadata_definitions]

    return hierarchy_dict


def extract_hierarchy_tojson(root, file_path):
    hierarchy_dict = get_hierarchy_dict(root)
    with open(file_path, 'w') as file:
        json.dump(hierarchy_dict, file)

    print('\nDone: structure tree exported at ' + file_path)

File: csharp/Windows/shared/shared_utils/PixyzUtils.cs
namespace PiXYZDemo
{
    using Newtonsoft.Json;
    using UnityEngine.Pixyz.API;
    using UnityEngine.Pixyz.Geom;
    using UnityEngine.Pixyz.Scene;
    using UnityEngine.Pixyz.View;


    public enum Orientation
    {
        FRONT = 0,
        BACK = 1,
        TOP = 2,
        BOTTOM = 3,
        LEFT = 4,
        RIGHT = 5
    }

    public static class PixyzUtils
    {
        public static (double, int, int, int) GetStats(uint root)
        {
            PixyzInit.api.Core.ConfigureInterfaceLogger(false, false, false);

            double t = DateTime.Now.TimeOfDay.TotalSeconds;
            var occurrences = new OccurrenceList((int)root);

            int n_triangles = (int)PixyzInit.api.Scene.GetPolygonCount(occurrences, true, false, false);
            int n_vertices = (int)PixyzInit.api.Scene.GetVertexCount(occurrences, false, false, false);
            int n_parts = PixyzInit.api.Scene.GetPartOccurrences((uint)root).length;
            PixyzInit.api.Core.ConfigureInterfaceLogger(true, true, true);

            return (t, n_triangles, n_vertices, n_parts);
        }

        public static void SaveScreenshot(int occurrence, string path, Orientation orientation = Orientation.FRONT,
                                   CameraType type = CameraType.Perspective, int resolution = 1024,
                                   int fov = 60, bool showEdges = false, bool showLines = false)
        {
            Console.WriteLine("Taking Screenshot...");

            Point3 direction = orientation switch
            {
                Orientation.FRONT => new Point3 { x = 0, y = 0, z = -1 },
                Orientation.BACK => new Point3 { x = 0, y = 0, z = 1 },
                Orientation.TOP => new Point3 { x = 0, y = -1, z = 0 },
                Orientation.BOTTOM => new Point3 { x = 0, y = 1, z = 0 },
                Orientation.LEFT => new Point3 { x = 1, y = 0, z = 0 },
                Orientation.RIGHT => new Point3 { x = -1, y = 0, z = 0 },
                _ => new Point3 { x = 0, y = 0, z = 0 }
            };

            var occurrences = new OccurrenceList(occurrence);

            var viewer = PixyzInit.api.View.CreateViewer(resolution, resolution);
            var gpuScene = PixyzInit.api.View.CreateGPUScene(occurrences, showEdges);
            PixyzInit.api.View.AddGPUScene(gpuScene, viewer);
            PixyzInit.api.View.FitCamera(direction, type, fov, viewer, occurrences);

            PixyzInit.api.View.SetViewerProperty("ShowEdges", showEdges ? "True" : "False", viewer);
            PixyzInit.api.View.SetViewerProperty("ShowLines", showLines ? "True" : "False", viewer);

            PixyzInit.api.View.TakeScreenshot(path, viewer);
            Console.WriteLine("Screenshot saved at: " + path);
            PixyzInit.api.View.DestroyViewer(viewer);
            PixyzInit.api.View.DestroyGPUScene(gpuScene);
        }

        public static void PrintStats(string fileName, double t, int n_triangles, int _n_triangles,
                               int n_vertices, int _n_vertices, int n_parts, int _n_parts)
        {
            Console.WriteLine("\n");
            Console.WriteLine($"{"file",-20}{fileName}\n");
            Console.WriteLine($"{"optimization",-20}{t:0.000} s\n");
            Console.WriteLine($"{"triangles",-20}{n_triangles} -> {_n_triangles}\n");
            Console.WriteLine($"{"vertices",-20}{n_vertices} -> {_n_vertices}\n");
            Console.WriteLine($"{"parts",-20}{n_parts} -> {_n_parts}\n");
        }

        static Dictionary<string, string> GetMetadataDict(PropertyValue propertyValue)
        {
            return new Dictionary<string, string>
            {
                { "name", propertyValue.name },
                { "value", propertyValue.value }
            };
        }

        static Dictionary<string, object> GetHierarchyDict(uint occurrence)
        {
            var hierarchyDict = new Dictionary<string, object>
            {
                { "name", PixyzInit.api.Core.GetProperty(occurrence, "Name") }
            };

            var matrix = PixyzInit.api.Scene.GetLocalMatrix(occurrence);
            var vector3lst = PixyzInit.api.Geom.ToTRS(matrix);
            hierarchyDict.Add("translation", new List<double> { vector3lst[0]._base.x, vector3lst[0]._base.y, vector3lst[0]._base.z });
            hierarchyDict.Add("rotation", new List<double> { vector3lst[1]._base.x, vector3lst[1]._base.y, vector3lst[1]._base.z });
            hierarchyDict.Add("scale", new List<double> { vector3lst[2]._base.x, vector3lst[2]._base.y, vector3lst[2]._base.z });

            var children = new List<Dictionary<string, object>>();
            foreach (var child in PixyzInit.api.Scene.GetChildren(occurrence).list)
            {
                children.Add(GetHierarchyDict(child));
            }
            hierarchyDict.Add("children", children);

            if (PixyzInit.api.Scene.HasComponent(occurrence, ComponentType.Metadata))
            {

                var metadataComponent = PixyzInit.api.Scene.GetComponent(occurrence, ComponentType.Metadata);
                var occurrences = new MetadataList(1);
                occurrences[0] = metadataComponent;

                MetadataDefinitionList metadataDefinitions = PixyzInit.api.Scene.GetMetadatasDefinitions(occurrences);


                MetadataDefinition metadataDefinition = metadataDefinitions[0];

                var metadataList = new List<Dictionary<string, string>>();

                foreach (PropertyValue propertyValue in metadataDefinition._base.list)
                {
                    metadataList.Add(GetMetadataDict(propertyValue));
                }

                hierarchyDict.Add("metadata", metadataList);
            }

            return hierarchyDict;
        }

        public static void ExtractHierarchyToJson(uint root, string filePath)
        {
            var hierarchyDict = GetHierarchyDict(root);
            using (var writer = new StreamWriter(filePath))
            {
                var json = JsonConvert.SerializeObject(hierarchyDict, Formatting.Indented);
                writer.Write(json);
            }
            Console.WriteLine("\nDone: structure tree exported at " + filePath);
        }
    }
}

